x-pack/platform/plugins/shared/observability_ai_assistant/server/service/distributed_lock_manager/index.ts

```diff
@@ -0,0 +1,329 @@
+   */
+  public async acquire(options: AcquireOptions = {}): Promise<boolean> {
+    await createLocksWriteIndexOnce(this.esClient);
+
+    this.token = uuid();
+    const ttl = options.ttlMs ?? 6 * 5 * 1000; // Default TTL: 5 minutes
+
+    this.logger.debug(
+      `Acquiring lock "${this.lockId}" with ttl = ${prettyMilliseconds(ttl)} and token = ${
+        this.token
+      }`
+    );
+
+    try {
+      const response = await this.esClient.update<LockDocument>({
+        index: LOCKS_CONCRETE_INDEX_NAME,
+        id: this.lockId,
+        scripted_upsert: true,
+        script: {
+          lang: 'painless',
+          source: `
+              // Get the current time on the ES server.
+              long now = System.currentTimeMillis();
+              
+              // If creating the document or if the lock is expired, update it.
+              if (ctx.op == 'create' || Instant.parse(ctx._source.expiresAt).toEpochMilli() < now) {
+                ctx._source.createdAt = Instant.ofEpochMilli(now).toString();
+                ctx._source.expiresAt = Instant.ofEpochMilli(now + params.ttl).toString();
+                ctx._source.metadata = params.metadata;
+                ctx._source.token = params.token;
+              } else {
+                ctx.op = 'noop'
+              }
+            `,
+          params: {
+            token: this.token,
+            ttl,
+            metadata: options.metadata ?? {},
+          },
+        },
+        // @ts-expect-error
+        upsert: {},
+        refresh: true,
+      });
+
+      if (response.result === 'created') {
+        this.logger.debug(
+          `Lock "${this.lockId}" acquired with ttl = ${prettyMilliseconds(ttl)} and token = ${
+            this.token
+          }`
+        );
+        return true;
+      } else if (response.result === 'updated') {
+        this.logger.debug(
+          ` Lock "${this.lockId}" was expired and re-acquired with ttl = ${prettyMilliseconds(
+            ttl
+          )} and token = ${this.token}`
+        );
+        return true;
+      } else if (response.result === 'noop') {
+        this.logger.debug(
+          `Lock "${this.lockId}" could not be acquired with token ${this.token} because it is already held`
+        );
+        return false;
+      } else {
+        throw new Error(`Unexpected response: ${response.result}`);
+      }
+    } catch (e) {
+      if (
+        e instanceof errors.ResponseError &&
+        e.body?.error?.type === 'version_conflict_engine_exception'
+      ) {
+        this.logger.debug(`Lock "${this.lockId}" already held (version conflict)`);
+        return false;
+      }
+
+      this.logger.error(`Failed to acquire lock "${this.lockId}": ${e.message}`);
+      this.logger.debug(e);
+      return false;
+    }
+  }
+
+  /**
+   * Releases the lock by deleting the document with the given lockId and token
+   */
+  public async release(): Promise<boolean> {
+    try {
+      const res = await pRetry(
```

Comment:
why does this retry automatically, always?

Reply:
can we use a scripted upsert here as well? I'm honestly not sure if what happens if we set `ctx.op` to `delete`. I think it would be better because we can distinguish between the lock existing, but not being owned, and the lock not existing at all. If not this is probably our best option.

Reply:
Removed the retry and changed to scripted upsert. This works fine:

```
            if (ctx._source.token == params.token) {
              ctx.op = 'delete';
            } else {
              ctx.op = 'noop';
            }
```